acme4j-utils/src/test/java/org/shredzone/acme4j/util/CSRBuilderTest.java

```diff
@@ -111,6 +91,109 @@ public void testGenerate() throws IOException {
      */
     @Test
     public void testECCGenerate() throws IOException {
+        CSRBuilder builder = createBuilderWithValues();
+
+        builder.sign(testEcKey);
+
+        PKCS10CertificationRequest csr = builder.getCSR();
+        assertThat(csr).isNotNull();
+        assertThat(csr.getEncoded()).isEqualTo(builder.getEncoded());
+
+        csrTest(csr);
+        writerTest(builder);
+    }
+
+    /**
+     * Make sure an exception is thrown when no domain is set.
+     */
+    @Test
+    public void testNoDomain() throws IOException {
+        IllegalStateException ise = assertThrows(IllegalStateException.class, () -> {
+            CSRBuilder builder = new CSRBuilder();
+            builder.sign(testKey);
+        });
+        Assert.assertEquals("unexpected exception message", "No domain or IP address was set", ise.getMessage());
+    }
+
+    /**
+     * Make sure an exception is thrown when an unknown identifier type is used.
+     */
+    @Test
+    public void testUnknownType() {
+        IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> {
+            CSRBuilder builder = new CSRBuilder();
+            builder.addIdentifier(new Identifier("UnKnOwN", "123"));
+        });
+        Assert.assertEquals("unexpected exception message", "Unknown identifier type: UnKnOwN", iae.getMessage());
+    }
+
+    /**
+     * Make sure all getters will fail if the CSR is not signed.
+     */
+    @Test
+    public void testNoSign() throws IOException {
+        CSRBuilder builder = new CSRBuilder();
+        IllegalStateException ise;
+
+        ise = assertThrows(IllegalStateException.class, builder::getCSR, "getCSR()");
+        Assert.assertEquals("unexpected exception message", "sign CSR first", ise.getMessage());
+        ise = assertThrows(IllegalStateException.class, builder::getEncoded, "getEncoded()");
+        Assert.assertEquals("unexpected exception message", "sign CSR first", ise.getMessage());
+        ise = assertThrows(IllegalStateException.class, () -> {
+            try (StringWriter w = new StringWriter()) {
+                builder.write(w);
+            }
+        }, "write()");
+        assertEquals("unexpected exception message", "sign CSR first", ise.getMessage());
+    }
+    
+    @Test
+    public void testAddAttrValues() throws Exception {
+        CSRBuilder builder = new CSRBuilder();
+        Exception ex;
+        String invAttNameExMessage = "";
+        try {
+            X500Name.getDefaultStyle().attrNameToOID("UNKNOWNATT");
+            fail("exception expected");
+        }
+        catch(IllegalArgumentException iae) {
+            invAttNameExMessage = iae.getMessage();
+        }
```

Comment:
You could also use `assertThrows` here, and save the somewhat outdated `try`/`fail`/`catch` combination.